### PR TITLE
ADR-31 Direct Get

### PR DESCRIPTION
**Pull request by tbeets**
This is an implementation ADR for the Direct Get and Subject-Appended Direct Get features.

### CODE REVIEW

File: adr/ADR-31.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+When Allow Direct is true, each of the stream's servers will also subscribe to `$JS.API.DIRECT.GET.<stream>.>` with
+fixed queue group `_sys_`.  Requests to this API endpoint will be interpreted as a shortcut for `last_by_subj` request
+where `last_by_subj` is derived by the token (or series of tokens) following the stream name rather than request payload.
+
+It is an error (408) if a client calls Subject-Appended Direct Get and includes a request payload.
+
+#### Response Format 
+
+Direct Get may return an error code:
```
Comment (kozlovic) [inline adr/ADR-31.md:98]: Just maybe add that it is through the use of headers, not a typical JetStream response object.

File: adr/ADR-31.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+- `404` if the request is valid but no matching message found in stream 
+- `408` if the request is empty or invalid
+
+Otherwise, Direct Get reply contains the message along with the following message headers:
+- `Nats-Stream`: stream name
+- `Nats-Sequence`: message sequence number 
+- `Nats-Time-Stamp`: message publish timestamp
+- `Nats-Subject`: message subject
+- `Nats-Last-Sequence`: sequence number of previous message for the given subject (if any)
```
Comment (kozlovic) [inline adr/ADR-31.md:107]: I don't think that this applies to direct get (more for republish feature I think?)

**Review by kozlovic — CHANGES_REQUESTED** [review]
Would remove reference to `Nats-Last-Sequence` since this does not apply to direct get (it is for republish) and maybe add a comment that the possible error returned is through a nats header, not a API error type of thing.

File: adr/ADR-31.md (hunk 2 above)
Comment (aricart) [reply adr/ADR-31.md:107]: The direct headers are these:
```
export enum DirectMsgHeaders {
// the stream where the message is stored
  Stream = "Nats-Stream",
// the sequence number of the stored message
  Sequence = "Nats-Sequence",
// the timestamp of the stored message
  TimeStamp = "Nats-Time-Stamp",
// the subject that was ingested
  Subject = "Nats-Subject",
}
```

Comment (tbeets) [reply adr/ADR-31.md:107]: Thanks.  Will remove.  Copy/paste from Issue 115 that did seem kind of awkward for this API.

File: adr/ADR-31.md (hunk 1 above)
Comment (tbeets) [reply adr/ADR-31.md:98]: Yes, I show in an example but good suggestion to clarify here.  Will add.

**Pull request merged**